vectortile/src/main/scala/geotrellis/vectortile/protobuf/ProtobufTile.scala

```diff
@@ -0,0 +1,293 @@
+  * A [[Layer]] decoded from Protobuf data. All of its Features are decoded
+  * lazily, making for very fast extraction of single features/geometries.
+  *
+  */
+// TODO The Layer storing `key` and `layout` is likely very wasteful
+// it terms of memory.
+case class ProtobufLayer(
+  private val rawLayer: vt.Tile.Layer,
+  private val tileExtent: Extent
+) extends Layer {
+  /* Expected fields */
+  def name: String = rawLayer.name
+  def extent: Int = rawLayer.extent.getOrElse(4096)
+
+  /** The version of the specification that this Layer adheres to. */
+  def version: Int = rawLayer.version
+
+  def features: Seq[Feature[Geometry, Map[String, Value]]] = {
+    points.append(multiPoints)
+     .append(lines)
+     .append(multiLines)
+     .append(polygons)
+     .append(multiPolygons)
+  }
+
+  /* Unconsumed raw Features */
+  // TODO Should this be a def?
+  private val (pointFs, lineFs, polyFs) = segregate(rawLayer.features)
+
+  /** How much of the [[Extent]] is covered by a single grid coordinate? */
+  private def resolution: Double = tileExtent.height / extent
+
+  /**
+   * Polymorphically generate a [[Stream]] of parsed Geometries and
+   * their metadata.
+   */
+  private def geomStream[G1 <: Geometry, G2 <: MultiGeometry](
+    feats: ListBuffer[vt.Tile.Feature]
+  )(implicit protobufGeom: ProtobufGeom[G1, G2]): Stream[(Either[G1, G2], Map[String, Value])] = {
+    def loop(fs: ListBuffer[vt.Tile.Feature]): Stream[(Either[G1, G2], Map[String, Value])] = {
+      if (fs.isEmpty) {
+        Stream.empty[(Either[G1, G2], Map[String, Value])]
+      } else {
+        val geoms = fs.head.geometry
+        val g = protobufGeom.fromCommands(Command.commands(geoms), tileExtent.northWest, resolution)
+
+        (g, getMeta(rawLayer.keys, rawLayer.values, fs.head.tags)) #:: loop(fs.tail)
+      }
+    }
+
+    loop(feats)
+  }
+
+  /**
+   * Construct Feature-specific metadata from the key/value lists of
+   * the parent layer.
+   */
+  private def getMeta(keys: Seq[String], vals: Seq[vt.Tile.Value], tags: Seq[Int]): Map[String, Value] = {
+    val pairs = new ListBuffer[(String, Value)]
+    var i = 0
+
+    while (i < tags.length) {
+      val k: String = keys(tags(i))
+      val v: vt.Tile.Value = vals(tags(i + 1))
+
+      pairs.append(k -> v)
+
+      i += 2
+    }
+
+    pairs.toMap
+  }
+
+  /* Geometry Streams */
+  private lazy val pointStream = geomStream[Point, MultiPoint](pointFs)
+  private lazy val lineStream = geomStream[Line, MultiLine](lineFs)
+  private lazy val polyStream = geomStream[Polygon, MultiPolygon](polyFs)
+
+  /* OPTIMIZATION NOTES
+   * `Stream.flatMap` maintains laziness. A common pattern here to "fold away"
+   * results you don't want are to use [[Option]]. However, flatMap here
+   * expects an [[Iterable]], and employs an implicit conversion from [[Option]]
+   * to get it.
+   *
+   * By calling directly what that implicit eventually calls at the bottom of
+   * its call stack, we save some operations.
```

Comment:
Haha because otherwise `case (Left(p), meta) => new ::(Feature(p, meta), Nil)` is some serious witchcraft.